scheduler/src/cook/kubernetes/api.clj

```diff
@@ -929,6 +929,22 @@
       ; synthetic pods (which don't specify a node name), but it
       ; doesn't hurt to add it for all pods we submit.
       (add-node-selector pod-spec cook-pool-label pool-name))
+
+    (when pod-constraints
+      (doseq [{:keys [constraint/attribute
+                      constraint/operator
+                      constraint/pattern]}
+              pod-constraints]
+        (condp = operator
```

Comment:
Shouldn't we be rejecting bad constraints at submit time instead of here, where they'll be invisibly ignored? Also, do we need/want any nil protection here?

Reply:
Yep, will take a look

Reply:
The bad constraint operators are already rejected at submission time:

```
$ echo '{"command": "true", "constraints": [["foo", "BOGUS", "bar"]]}' | cs submit --raw
Enter the raw job(s) JSON (press Ctrl+D on a blank line to submit)
Attempting to submit on dev0 cluster...
Job submission failed on dev0:
{"jobs": [{"constraints": [[null, "(named (not (constraint-operator-exists? \"BOGUS\")) \"operator\")", null]]}]}

Job submission failed on all of your configured clusters.
```

Reply:
Changed the log line to a `throw` so a bad constraint operator (which is already blocked at submission time) would be more visible -- the pod would not get submitted.

Reply:
`null` attributes and patterns are also blocked at submission time:

```
$ echo '{"command": "true", "constraints": [[null, "EQUALS", null]]}' | cs submit --raw                                                                                                                                                          
Enter the raw job(s) JSON (press Ctrl+D on a blank line to submit)
Attempting to submit on dev0 cluster...
Job submission failed on dev0:
{"jobs": [{"constraints": [["(named (not (instance? java.lang.String nil)) \"attribute\")", null, "(named (not (instance? java.lang.String nil)) \"pattern\")"]]}]}

Job submission failed on all of your configured clusters.
```

Reply:
Looks good to me, but I'd put in the throw string (or an adjacent comment) that these invalid situations should have been rejected at submit time.

Reply:
Good call, added the comment below

```
          ; A bad constraint operator should have
          ; been blocked at job submission time
```